---
title: "Re-analysis of GSE21933 dataset: Understanding Phalanx Human OneArray Data"
author: "Spandan Pandya"
date: "2025-01-24"
output:
  html_document:
    df_print: paged
editor_options: 
  markdown: 
    wrap: 72
---

# About the data:

Series ID:
[GSE21933](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=gse21933)
Platform: [Phalanx](http://www.phalanxbiotech.com/) Human OneAray
[GPL6254](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GPL6254)

Experiment type: Expression profiling by Array Details:

Phalanx Biotech Group's Human OneArray contains 32,048 features, 30968
detection probes and 1080 control probes, spotted onto glass slides
using a proprietary non-contact printing method. Detection probes are
annotated against the human genome and grouped into the following
categories:

-   Group 1 - gene specific: exon

-   Group 2 - intron hit

-   Group 3 - intergenic

-   Group 4 - multi-gene hits

-   Group 5 - no hit to genome

-   Group 6 - \>200hits to genome (Mostly represents control sets)

Lo FY, Chang JW, Chang IS, Chen YJ et al. The database of chromosome
imbalance regions and genes resided in lung cancer from Asian and
Caucasian identified by array-comparative genomic hybridization. BMC
Cancer 2012 Jun 12;12:235. PMID: 22691236

```{=html}
<!-- This HTML is for increasing and adjusting the size of the html output -->
<style type="text/css">
.main-container {
  max-width: 1800px;
  margin-left: auto;
  margin-right: auto;
}
</style>
```

```{r echo=FALSE, message=FALSE, warning=FALSE}
library(DESeq2)
library(limma)
library(edgeR)
library(GEOquery)
library(pheatmap)
library(org.Hs.eg.db)
library(ggplot2)
library(ggrepel)
library(tidyr)
library(tidyverse)
library(stats)
```

## Data and Normalization

```{r}
data_id <- "GSE21933"

eset <-getGEO(data_id)
gse <- eset[[1]]

head(pData(gse)) #phenotype data
head(fData(gse)) #feature data
head(exprs(gse)) #complete expression data set

summary(exprs(gse)) #statistical analysis of all summary()

```

The Phenotype Data contains 42 samples and 40 features for the samples.
The Feature Data: The feature data contains 30967 probes and 10 features
for each samples.

Looking at the summary of expressions, the values for each sample falls
between 0-16. Suggesting that these values a log2 normalized. Further,
the figure below shows a boxplot of all the gene count values for each
sample. As seen below, all the samples have lined up unifromly
suggesting that the data is already log normalized.

```{r}
boxplot(exprs(gse), outline = FALSE, col = "gold")
```

# Inspecting clinical Data

There are several columns within the metadata many of these are
repeating. The most important of these columns are the column-1 -
`title`, and columns 36 to 40
`age:ch1`,`histology:ch1`,`sex:ch1`,`stage:ch1`,`tissue:ch1`. Further,
most of the column names have a ":ch1" at the end. This was removed.
Lastly, the histology and stage columns have no data for negative
samples. This was changed to "neg".

```{r}
#Preparing sample metadata
samplesinfo <- pData(gse)
samplesinfo <- samplesinfo[,c(1,36:40)]
colnames(samplesinfo) <- gsub(":ch1","",colnames(samplesinfo))

samplesinfo <- samplesinfo%>%
  mutate(across(c(3,5),~replace_na(.,"Neg")))

samplesinfo <- samplesinfo%>%
  mutate_all(~gsub(" years","",.))%>%
  mutate(across(c(histology,sex,stage,tissue),factor))%>%
  mutate(age = as.numeric(age))

#Prepare Features Metadata 
featuresinfo <- fData(gse)
#Prepare expression data
exprs_data <- exprs(gse)
```

The data set represents 42 lung tissue samples, 21 primary normal lung
tissues and 21 primary lung tumor tissues. The tumor tissues are
representing six stages of lung cancer IA,IB,IIB,IIIA,IIIB,IV. Further
two hematological patterns are also represented in tumor samples, these
are: adenocarcinoma (AD) and Squamous cell carcinoma (SQ).

```{r}
samplesinfo%>%
  group_by(tissue, stage,histology)%>%
  tally()%>%
  spread(histology,n)
```

# Clustering and PCA analysis

```{r fig.height=15, fig.width=20}
corMatrix <- cor(exprs_data, use = "c")
rownames(samplesinfo) <- colnames(corMatrix)
pheatmap(corMatrix, annotation_col =samplesinfo[,3:6], annotation_row = samplesinfo[,3:6], cluster_rows = T, cluster_cols = T)

```

```{r fig.height=15, fig.width=20}
pca_gse <- prcomp(t(exprs_data)) # Run PCA analysis

screeplot(pca_gse, npcs=min(10,length(pca_gse$sdev)),type = c("barplot","lines"))

cbind(samplesinfo, pca_gse$x)%>%
  ggplot(aes(x= PC1, y = PC2, col=tissue, label = paste(tissue)))+
  geom_point()+
  geom_text_repel()

cbind(samplesinfo, pca_gse$x)%>%
  ggplot(aes(x= PC1, y = PC2, col=stage, label = paste(stage)))+
  geom_point()+
  geom_text_repel()

cbind(samplesinfo, pca_gse$x)%>%
  ggplot(aes(x= PC1, y = PC2, col=histology, label = paste(histology)))+
  geom_point()+
  geom_text_repel()

cbind(samplesinfo, pca_gse$x)%>%
  ggplot(aes(x= PC1, y = PC2, col=title, label = paste(title)))+
  geom_point(size = 5)+
  geom_text_repel(size = 5)

```

# Filter dataset:

Lowly expressed genes must be filtered out as they can deviate the
results of gene expression. Here I am applying three separate
strategies: All samples have a data, at least 50 percent have a data

```{r}
cutoff <- median(exprs_data) #take median of expression dataset
is_expressed <- exprs_data>cutoff
keep_all<- rowSums(is_expressed) >=42 #stringent should be present in all samples.
keep_50 <- rowSums(is_expressed)>=21 #should be present in atleasst 50% samples.
keep <- rowSums(is_expressed)>2 

gse_filt <- rbind(table(keep),table(keep_50),table(keep_all))
rownames(gse_filt) <- c("2 or more","50%","100%")
gse_filt

gse__2 <- gse[keep,]
gse_50 <- gse[keep_50,]
gse_all <- gse[keep_all,]
```

## Filter genes:

# Differential Expression

## Combined effect and interaction effect

```{r}
set.seed(1234)
design_gse_combined <- model.matrix(~0+factor(samplesinfo$tissue)+factor(samplesinfo$histology)+factor(samplesinfo$stage))
design_gse_interaction <- model.matrix(~0+factor(samplesinfo$tissue)*factor(samplesinfo$histology)*factor(samplesinfo$stage))
colnames(design_gse_combined)<-  c("tumor","normal","Neg","SQ","IB","IIB","IIIA","IIIB","IV","Neg") 

fit <- lmFit(exprs(gse_all),design_gse_combined) #fit all the genes
contrasts_gse <- makeContrasts(tumor-normal,levels =design_gse_combined)
fit2<-contrasts.fit(fit,contrasts_gse)%>%
  eBayes()
head(topTable(fit2))
table(decideTests(fit2))

full_results_all <- topTable(fit2,number = Inf)
ggplot(full_results_all,aes(x=logFC, y = B))+geom_point()


#ggplot(full_results_all, aes(x = logFC, y = -log10(adj.P.Val),)) +
#  geom_point(aes(color = adj.P.Val < 0.05), alpha = 0.5) +
#  scale_color_manual(values = c("black", "red")) +
#  labs(title = paste("Volcano Plot:", ), x = "Log2 Fold Change", y = "-Log10 Adjusted P-value") +
#  theme_minimal()
```

## AD versus SQ versus neg

```{r}
design_gse <- model.matrix(~0+samplesinfo$histology)
colnames(design_gse)<-c("AD","neg","SQ")

fit <- lmFit(exprs(gse_all),design_gse) #fit all the genes
contrasts_gse <- makeContrasts( ADvsneg = AD-neg, #AD versus negative
                                SQvsneg = SQ-neg, #SQ versus negative
                                ADvsSQ = AD-SQ, #AD versus SQ
                                SQvsAD = SQ-AD, #SQ versus AD
                                avsavg = (AD+SQ)/2-neg, #negative versus combined disease how significantly different SQ and AD are from negative.
                                ADvsSQtoneg = ((AD-neg)-(SQ-neg)), #Checks if AD is significantly more different than SQ
                                levels =design_gse)
cont_names <- colnames(contrasts_gse)

result_all <- data.frame()
for (i in cont_names){
  fit_contrast <- contrasts.fit(fit, contrasts_gse[,i])
  fit_contrast <- eBayes(fit_contrast)
  results <- topTable(fit_contrast, adjust="fdr", number=Inf)
  results$Genes <- rownames(results)
  results$Contrasts<- i
  result_all <- bind_rows(result_all,results)
  
  plot <-ggplot(results, aes(x = logFC, y = -log10(adj.P.Val))) +
    geom_point(aes(color = adj.P.Val < 0.05), alpha = 0.5) +
    scale_color_manual(values = c("black", "red")) +
    labs(title = paste("Volcano Plot:", i), x = "Log2 Fold Change", y = "-Log10 Adjusted P-value") +
    theme_minimal()
  print(plot)
  ggsave(filename = paste0(i,"_spandan.png"))
  
}

```

```{r}
#volcano plot
ggplot(result_all, aes(x = logFC, y = -log10(adj.P.Val), color = Contrasts)) +
  geom_point(alpha = 0.5) +
  labs(title = "Volcano Plot for Multiple Contrasts", x = "Log2 Fold Change", y = "-Log10 Adjusted P-value") +
  theme_minimal()

```

Reference: tutorial:
"<https://sbc.shef.ac.uk/geo_tutorial/tutorial.nb.html>"
