---
title: "Re-analysis of GSE21933 dataset: Understanding Phalanx Human OneArray Data"
author: "Spandan Pandya"
date: "2025-01-24"
output:
  html_document:
    df_print: paged
editor_options: 
  markdown: 
    wrap: 72
---

# About the data:

Series ID:
[GSE21933](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=gse21933)
Platform: [Phalanx](http://www.phalanxbiotech.com/) Human OneAray v5
[GPL6254](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GPL6254)

Experiment type: Expression profiling by Array Details:

Phalanx Biotech Group's Human OneArray v5 contains 32,048 features, 30968
detection probes and 1080 control probes, spotted onto glass slides
using a proprietary non-contact printing method. Detection probes are
annotated against the human genome and grouped into the following
categories:

-   Group 1 - gene specific: exon

-   Group 2 - intron hit

-   Group 3 - intergenic

-   Group 4 - multi-gene hits

-   Group 5 - no hit to genome

-   Group 6 - \>200hits to genome (Mostly represents control sets)

Lo FY, Chang JW, Chang IS, Chen YJ et al. The database of chromosome
imbalance regions and genes resided in lung cancer from Asian and
Caucasian identified by array-comparative genomic hybridization. BMC
Cancer 2012 Jun 12;12:235. PMID: 22691236


```{=html}
<!-- This HTML is for increasing and adjusting the size of the html output -->
<style type="text/css">
.main-container {
  max-width: 1800px;
  margin-left: auto;
  margin-right: auto;
}
</style>
```

```{r library_Prep, echo=FALSE, message=FALSE, warning=FALSE}
install.packages("tidyverse")
install.packages("ggpubr")
install.packages("reshape2")
library(readr)
library(DESeq2)
library(limma)
library(edgeR)
library(GEOquery)
library(pheatmap)
library(org.Hs.eg.db)
library(ggplot2)
library(ggrepel)
library(tidyr)
library(tidyverse)
library(stats)
library(ggsci)
library(rmarkdown)
library(knitr)
library(ggpubr)
library(WGCNA)
```

## Data and Normalization

```{r Acquire_Data}
data_id <- "GSE21933"

eset <-getGEO(data_id) #get dataset
gse <- eset[[1]] #get gene dataset

# head(pData(gse)) #phenotype data
# head(fData(gse)) #feature data
# head(exprs(gse)) #complete expression data set

summary(exprs(gse)) #statistical analysis of all summary()

boxplot(exprs(gse), outline = FALSE, col = "gold") #Visualize the normalization

```

The Phenotype Data contains 42 samples and 40 features for the samples.
The Feature Data: The feature data contains 30967 probes and 10 features
for each samples.

Looking at the summary of expressions, the values for each sample falls
between 0-16. Suggesting that these values a log2 normalized. Further,
the figure below shows a boxplot of all the gene count values for each
sample. As seen below, all the samples have lined up unifromly
suggesting that the data is already log normalized.

# Inspecting clinical Data
There are several columns within the metadata many of these are
repeating. The most important of these columns are the column-1 -
`title`, and columns 36 to 40
`age:ch1`,`histology:ch1`,`sex:ch1`,`stage:ch1`,`tissue:ch1`. Further,
most of the column names have a ":ch1" at the end. This was removed.
Lastly, the histology and stage columns have no data for negative
samples. This was changed to "neg".

```{r sample_metadat}
#Preparing sample metadata
samplesinfo <- pData(gse) #gettign sample data
samplesinfo <- samplesinfo[,c(1,36:40)] #There are multiple columns for same data, acquiring relevant metadata information.
colnames(samplesinfo) <- gsub(":ch1","",colnames(samplesinfo))

samplesinfo <- samplesinfo%>%
  mutate(across(c(3,5),~replace_na(.,"Neg")))

samplesinfo <- samplesinfo%>%
  mutate_all(~gsub(" years","",.))%>%
  mutate(across(c(histology,sex,stage,tissue),factor))%>%
  mutate(age = as.numeric(age))%>%
  mutate(tissue = ifelse(tissue == "primary normal lung tissue","normal","tumor"))

samplesinfo$histology <- relevel(samplesinfo$histology, ref = "Neg")
samplesinfo$stage <- relevel(samplesinfo$stage, ref = "Neg")
samplesinfo$tissue <- relevel(as.factor(samplesinfo$tissue), ref = "normal")

#Prepare Features Metadata 
featuresinfo <- fData(gse)
#Prepare expression data
exprs_data <- exprs(gse)

samplesinfo%>%
  group_by(tissue, stage,histology)%>%
  tally()%>%
  spread(histology,n)

```

The data set represents 42 lung tissue samples, 21 primary normal lung
tissues and 21 primary lung tumor tissues. The tumor tissues are
representing six stages of lung cancer IA,IB,IIB,IIIA,IIIB,IV. Further
two hematological patterns are also represented in tumor samples, these
are: adenocarcinoma (AD) and Squamous cell carcinoma (SQ).

# Clustering and PCA analysis

```{r correlation, fig.height=10, fig.width=12}
corMatrix <- cor(exprs_data, use = "c")
rownames(samplesinfo) <- colnames(corMatrix)
pheatmap(corMatrix, annotation_col =samplesinfo[,3:6], annotation_row = samplesinfo[,3:6], cluster_rows = T, cluster_cols = T)

```

```{r PCA, fig.height=10, fig.width=20}

#Hierarchical cluster based dimension analysis
htree <- hclust(dist(t(exprs_data)), method = "average")
plot(htree) #It seems there is a specific pattern of clustering within samples. And this might be biologically relevant difference so I think it is better to makes sure we check with PCA,the variance cotnribution. 


pca_gse <- prcomp(t(exprs_data)) # Run PCA analysis

screeplot(pca_gse, npcs=min(10,length(pca_gse$sdev)),type = c("barplot","lines")) #Screeplots of all the PC;s and their cotnribution. 

# PCA plot by tissue
p1 <- cbind(samplesinfo, pca_gse$x)%>%
  ggplot(aes(x= PC1, y = PC2, col=tissue, label = paste(tissue)))+
  geom_point(size=5)+
  geom_text_repel()
  

p2 <- cbind(samplesinfo, pca_gse$x)%>%
  ggplot(aes(x= PC1, y = PC2, col=stage, label = paste(stage)))+
  geom_point()+
  geom_text_repel()

p3 <- cbind(samplesinfo, pca_gse$x)%>%
  ggplot(aes(x= PC1, y = PC2, col=histology, label = paste(histology)))+
  geom_point()+
  geom_text_repel()

p4 <- cbind(samplesinfo, pca_gse$x)%>%
  ggplot(aes(x= PC1, y = PC2, col=title, label = paste(title)))+
  geom_point(size = 5)+
  geom_text_repel(size = 5)+
  ggsci::scale_fill_nejm()


ggarrange(p1,p2,p3,p4, ncol = 2, nrow = 2, common.legend = TRUE, legend = "bottom")

```

# Filter dataset:

First I am filtering out only genes that belong to group-6 in the array as that represents the control genes. Lowly expressed genes must be filtered out as they can deviate the results of gene expression. The cut off for low expression is median gene expression of each sample. Next, I am applying three seperate strategies: 1) All samples have a data
                                       2) At least 50%  samples have all the data
                                       3) At least 25% of sample have a data.
However, this is a microarray dataset. 

## Exploring cut offs: 
```{r control_identification}

control_list <- rownames(gse@featureData[which(gse@featureData@data$GROUP ==6)]) #create a vector list of controls
control_matrix <- exprs_data[rownames(exprs_data)%in% control_list,] #isolate a matrix of just controls

mean_control <- colMeans(control_matrix)
mean_control <- data.frame(mean_control)%>%
  rownames_to_column(var = "control")
colnames(mean_control) <- c("controls","mean")

cat(paste0("Mean of Controls= ",mean(control_matrix),"\n","Mean of data= ",mean(exprs_data),"\n", "Median of Controls= ",median(control_matrix),"\n","Median of data= ",median(exprs_data)))
```

```{r filter_based_on_samples}
cutoff <- median(exprs_data) #take median of expression dataset
is_expressed <- exprs_data>cutoff

keep_all<- rowSums(is_expressed) >=42 #stringent should be present in all samples.
keep_50 <- rowSums(is_expressed)>=21 #should be present in at leasst 50% samples.
keep <- rowSums(is_expressed)<10 # lenient threshold present in at least 25% samples

gse_filt <- rbind(table(keep),table(keep_50),table(keep_all))
rownames(gse_filt) <- c("25%","50%","100%")
gse_filt

gse_25 <- gse[keep,]
gse_50 <- gse[keep_50,]
gse_100 <- gse[keep_all,]

dim_table <- data.frame(SampleCount = c("25%","50%","100%"),
                        features = c(dim(gse_25)[1],dim(gse_50)[1],dim(gse_100)[1]),
                        features = c(dim(gse_25)[2],dim(gse_50)[2],dim(gse_100)[2]))

print(dim_table)



plot(density(control_matrix), main = "Expression Density", xlab = "log2 intensity",)
abline(v = c(1,6), col = c("red","blue"), lty = 2)  # typical cutoff


```

## Filter genes:
```{r filter_exons}
#Filter only sampels that are group-1 or known genes 
exon_genes <- row.names(featuresinfo[which(featuresinfo$GROUP == 1),])
gse[exon_genes,]
exprs_data <- exprs_data[exon_genes,]
```

# Differential Expression

## Combined effect and interaction effect

```{r design_matrix}
set.seed(1234)
design_combined <- model.matrix(~0+tissue+ histology+stage, data = samplesinfo)
design_interaction <- model.matrix(~tissue*histology*stage, data = samplesinfo)
```

## AD versus SQ versus neg

```{r ADvsSQ}
design_gse <- model.matrix(~0+samplesinfo$histology)
colnames(design_gse)<-c("neg","AD","SQ")

fit <- lmFit(exprs(gse_100),design_gse) #fit all the genes
contrasts_gse <- makeContrasts( ADvsneg = AD-neg, #AD versus negative
                                SQvsneg = SQ-neg, #SQ versus negative
                                ADvsSQ = AD-SQ, #AD versus SQ
                                SQvsAD = SQ-AD, #SQ versus AD
                                avsavg = (AD+SQ)/2-neg, #negative versus combined disease how significantly different SQ and AD are from negative.
                                ADvsSQtoneg = ((AD-neg)-(SQ-neg)), #Checks if AD is significantly more different than SQ
                                levels =design_gse)
cont_names <- colnames(contrasts_gse)

result_all <- data.frame()
for (i in cont_names){
  fit_contrast <- contrasts.fit(fit, contrasts_gse[,i])
  fit_contrast <- eBayes(fit_contrast)
  results <- topTable(fit_contrast, adjust="fdr", number=Inf)
  results$Genes <- rownames(results)
  results$Contrasts<- i
  result_all <- bind_rows(result_all,results)
  
  plot <-ggplot(results, aes(x = logFC, y = -log10(adj.P.Val))) +
    geom_point(aes(color = adj.P.Val < 0.05), alpha = 0.5) +
    scale_color_manual(values = c("black", "red")) +
    labs(title = paste("Volcano Plot:", i), x = "Log2 Fold Change", y = "-Log10 Adjusted P-value") +
    theme_minimal()
  print(plot)
  ggsave(filename = paste0(i,"_volcano.png"), path = "../figures")
  
}

```

```{r volcano_plot}
#volcano plot
ggplot(result_all, aes(x = logFC, y = -log10(adj.P.Val), color = Contrasts)) +
  geom_point(alpha = 0.5) +
  labs(title = "Volcano Plot for Multiple Contrasts", x = "Log2 Fold Change", y = "-Log10 Adjusted P-value") +
  theme_minimal()
ggsave(filename = "volcano_all_contrasts.png",path = "../figures")

```

# WGCNA


```{r good_samples, eval=FALSE, include=FALSE}
gsg <- goodSamplesGenes(t(exprs_data)) #explore data for WGCNA to ensure all genes and columns are good. 

print(paste0("All samples are looking good: "))
head(gsg$allOK) #IF TRUE all the values are OK. 
head(gsg$goodGenes) #All genes are OK. 
head(gsg$goodSamples) # all samples are OK (42 of 42)
```

## Chose Threshold
```{r softpower_threshold, eval=FALSE, include=FALSE}
powers <- c(1:50)

sft <- pickSoftThreshold(t(exprs_data), powerVector = powers, verbose = 5) #analyze power at multiple threshold values

#plot to choose an optimal power

# Extract data

sft_df <- sft$fitIndices%>%
  select(Power = Power,
                SFT_R2 = SFT.R.sq,
                Slope = slope,
                MeanConnectivity = mean.k.)
sft_df <- sft_df%>%
  mutate(SignedR2 = -sign(Slope)*SFT_R2)


#Scale-Free Topology Index plot 
sfree <- ggplot(sft_df,aes(x = Power, y = SignedR2, label = Power))+
  geom_point(color= "#FF0099",size = 3)+
  geom_text(vjust = 1.5, size = 3.5)+
  geom_hline(yintercept = c(0.8,0.9), linetype= "dashed", color = c("#300666","#300999"),linewidth=1)+
  geom_vline(xintercept = 15, linetype= "dashed", color = "red",linewidth=0.6)+
  labs(title = "Scale-Free Topology Fit Index", y= bquote(paste("Signed ",R^2)))+
  theme_pubr()

#Mean connectivity plot

mfree <- ggplot(sft_df, aes(x = Power, y = MeanConnectivity, label=Power))+
  geom_point(color = "#FF0099", size = 3)+
  geom_text(vjust=-1, size=3.5)+
  labs(title = "Mean Connectivity", y = "Mean Connectivity", x = "Soft Threshold") +
  geom_hline(yintercept = 100, linetype= "dashed", color = "#300999",linewidth=1)+
  geom_vline(xintercept = 15, linetype= "dashed", color = "red",linewidth=0.6)+
  theme_pubr()

ggarrange(plotlist = list(sfree,mfree), nrow = 2,ncol = 1)
ggsave(filename="WGCNA_Threshold.png", path = "../figures", dpi = 600, width = 20, height = 10, units = "in" )

chosen_threshold <- sft_df$Power[12]
mean_conectivity <- sft_df$MeanConnectivity[chosen_threshold]
scale_free_top <- sft_df$SFT_R2[chosen_threshold]


```
I chose the threshold to be at 15 as this had a mean connectivity of  at scale free topology index of 


```{r WGCNA_analysis, fig.height=9, fig.width=12}
softPower <- 15

cat("The soft power is",softPower)

adj_matrix <- t(exprs_data) #filtered genes present in all samples
adjacency_mat <- adjacency(adj_matrix, power = softPower)

# Turn adjacency into topological overlap
TOM <- TOMsimilarity(adjacency_mat)
dissTOM <- 1 - TOM

# Hierarchical clustering
geneTree <- hclust(as.dist(dissTOM), method = "average")
plot(geneTree, main = "Gene clustering")

# Module identification using dynamic tree cut deepsplit allows adjusting granulatity. 
dynamicMods <- cutreeDynamic(dendro = geneTree, distM = dissTOM,
                             deepSplit = 2, pamRespectsDendro = FALSE,
                             minClusterSize = 100) #minClustersize seems optimal with at least 100 genes per cluster. 
cat("DYnamic tree with modules:")
table(dynamicMods)

# Assign module colors
dynamicColors <- labels2colors(dynamicMods)
plotDendroAndColors(geneTree, dynamicColors, "Dynamic Tree Cut",
                    dendroLabels = FALSE, hang = 0.03,
                    addGuide = TRUE, guideHang = 0.05)

cat("DYnamic tree modules with colors:")
table(dynamicColors)

# STEP-4: Relate Modules to Traits
# Calculate eigengenes
MEs <- moduleEigengenes(adj_matrix, colors = dynamicColors)$eigengenes
rowname_MEs <- row.names(MEs)
MEs <- orderMEs(MEs)
row.names(MEs) <- rowname_MEs

# Correlate eigengenes with traits
traits <- samplesinfo %>%
  mutate(across(where(is.factor), ~as.numeric(as.factor(.)))) %>%
  select(-c("title","sex"))  # Remove non-trait columns like ID if present

#traits <- model.matrix(~ histology + sex + stage + tissue + age, data = samplesinfo)[,-1]


moduleTraitCor <- cor(MEs,traits, use = "p")
moduleTraitPvalue <- corPvalueStudent(moduleTraitCor, nSamples = ncol(adj_matrix))

all(rownames(MEs) == rownames(traits))  # must be TRUE


# Plot correlation heatmap

# Load required libraries
library(reshape2)
# Prepare correlation and p-value matrices
moduleTraitCor <- cor(MEs, traits, use = "p")
moduleTraitPvalue <- corPvalueStudent(moduleTraitCor, nSamples = ncol(adj_matrix))

# Melt into long format for ggplot
df_cor <- melt(moduleTraitCor)
df_pval <- melt(moduleTraitPvalue)

# Combine and format labels
df_combined <- df_cor
df_combined$Pvalue <- df_pval$value
df_combined$Label <- paste0(signif(df_cor$value, 2), "\n(",
                            signif(df_pval$value, 1), ")")

# Optional: set factor levels to preserve order
df_combined$Var1 <- factor(df_combined$Var1, levels = rownames(moduleTraitCor))
df_combined$Var2 <- factor(df_combined$Var2, levels = colnames(moduleTraitCor))

# Plot
ggplot(df_combined, aes(x = Var2, y = Var1, fill = value)) +
  geom_tile(color = "grey90") +
  geom_text(aes(label = Label), size = 3.5) +
  scale_fill_gradient2(low = "blue", mid = "white", high = "red", midpoint = 0,
                       name = "Correlation") +
  theme_minimal(base_size = 14) +
  labs(x = "Traits", y = "Modules", title = "Module-Trait Relationships") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        panel.grid = element_blank(),
        plot.title = element_text(hjust = 0.5))
ggsave(filename = "WGCNA_ModuleCorr.png",path = "../figures")

names(dynamicColors) <- colnames(adj_matrix)
kWithin <- intramodularConnectivity(adjacency_mat, dynamicColors) # Calculate intra-modular connectivity (hubness)

library(readr)

# Function to check if a string is a single valid entry
is_single_valid <- function(x) {
  !is.na(x) & x != "" & !grepl(";", x)
}

for (m in unique(dynamicColors)) {
  module_genes <- names(dynamicColors)[which(dynamicColors == m)]
  
  if (length(module_genes) == 0) next
  
  # Order genes by intramodular connectivity
  hub_genes <- module_genes[order(kWithin[module_genes, "kWithin"], decreasing = TRUE)]
  
  # Extract columns
  gene_symbol <- featuresinfo[hub_genes, "GENE_SYMBOL"]
  ensembl_id <- featuresinfo[hub_genes, "ENSEMBL_GENE_ID"]
  
  # Determine final name using the priority rules
  final_names <- character(length(hub_genes))
  for (i in seq_along(hub_genes)) {
    if (is_single_valid(gene_symbol[i])) {
      final_names[i] <- gene_symbol[i]
    } else if (is_single_valid(ensembl_id[i])) {
      final_names[i] <- ensembl_id[i]
    } else {
      final_names[i] <- hub_genes[i]  # fallback to original name
    }
  }
  
  # Write final names to output
  write_lines(final_names, file = paste0("../outputs/hub_", m, ".txt"))
}

```
# tutorial:
https://sbc.shef.ac.uk/geo_tutorial/tutorial.nb.html
